clients/src/main/java/org/apache/kafka/common/ApiKey.java

```diff
@@ -0,0 +1,127 @@
+ * compatibility, we can add more APIs to the list, but not remove or reorder existing APIs.
+ */
+public enum ApiKey {
+    PRODUCE("Produce"),
+    FETCH("Fetch"),
+    LIST_OFFSETS("Offsets"),
+    METADATA("Metadata"),
+    LEADER_AND_ISR("LeaderAndIsr"),
+    STOP_REPLICA("StopReplica"),
+    UPDATE_METADATA_KEY("UpdateMetadata"),
+    CONTROLLED_SHUTDOWN_KEY("ControlledShutdown"),
+    OFFSET_COMMIT("OffsetCommit"),
+    OFFSET_FETCH("OffsetFetch"),
+    FIND_COORDINATOR("FindCoordinator"),
+    JOIN_GROUP("JoinGroup"),
+    HEARTBEAT("Heartbeat"),
+    LEAVE_GROUP("LeaveGroup"),
+    SYNC_GROUP("SyncGroup"),
+    DESCRIBE_GROUPS("DescribeGroups"),
+    LIST_GROUPS("ListGroups"),
+    SASL_HANDSHAKE("SaslHandshake"),
+    API_VERSIONS("ApiVersions"),
+    CREATE_TOPICS("CreateTopics"),
+    DELETE_TOPICS("DeleteTopics"),
+    DELETE_RECORDS("DeleteRecords"),
+    INIT_PRODUCER_ID("InitProducerId"),
+    OFFSET_FOR_LEADER_EPOCH("OffsetForLeaderEpoch"),
+    ADD_PARTITIONS_TO_TXN("AddPartitionsToTxn"),
+    ADD_OFFSETS_TO_TXN("AddOffsetsToTxn"),
+    END_TXN("EndTxn"),
+    WRITE_TXN_MARKERS("WriteTxnMarkers"),
+    TXN_OFFSET_COMMIT("TxnOffsetCommit"),
+    DESCRIBE_ACLS("DescribeAcls"),
+    CREATE_ACLS("CreateAcls"),
+    DELETE_ACLS("DeleteAcls"),
+    DESCRIBE_CONFIGS("DescribeConfigs"),
+    ALTER_CONFIGS("AlterConfigs");
+
+    public static final ApiKey[] VALUES = ApiKey.values();
+
+    public static final short NUM_KEYS = (short) VALUES.length;
+
+    private final String title;
+
+    public static boolean hasId(short id) {
+        return (id >= 0) && (id < NUM_KEYS);
+    }
+
+    public static ApiKey fromId(int id) {
+        if ((id < 0) || (id >= NUM_KEYS)) {
+            throw new IllegalArgumentException(String.format("Unexpected ApiKey id `%d`, " +
+                "it should be between 0 and %d (inclusive)", id, NUM_KEYS - 1));
+        }
+        return VALUES[id];
+    }
+
+    ApiKey(String title) {
+        this.title = title;
+    }
+
+    /**
+     * Get the numeric id fo this ApiKey.
+     *
+     * The numeric ID of an API is fixed.  It cannot be changed once the API is released.
+     */
+    public short id() {
+        return (short) ordinal();
+    }
+
+    /**
+     * Get the human-readable title of this ApiKey.
+     *
+     * The titles of an API can change over time.
+     */
+    public String title() {
+        return title;
+    }
+
+    /**
+     * Get the range of versions supported by this software.
+     *
+     * @throws UnsupportedVersionException       If the software does not support any version of this API.
+     * @return                                   The version range supported by this software,
+     *                                           which could be different than the software you are
+     *                                           communicating with over the network.
+     */
+    public ApiVersionRange supportedRange() {
```

Comment:
`ApiKey#supportedRange(ApiKey key)` returns the range of versions the current software supports for `key`.  `NodeVersions#apiVersionRange(ApiKey api)` returns the range of versions that the node supports for `key`.  These are going to be different, unless both broker and client are on the same software version.

> Also, it feels a bit weird for a public facing class to reference Protocol, which is not a public facing one.
The original patch had the supported versions coded in `ApiKey`, so that there was no reference to `Protocol`.  @ijuma objected to this since it meant that there was another place in the code that would have to be changed when creating a new version of an API.

I think both approaches are reasonable.  We could put the supported versions into ApiKey and eliminate the reference to protocol.  Then Protocol could just validate that the versions matched up with its versions.  Or, we could also do what we're doing now, which minimizes the work needed to bump an API version.  Since this detail is not exposed to users, we can change our approach later with no compatibility problems.